### PR TITLE
Add nested transaction support to index plugins (#361)

### LINKED ISSUE #361: Replayable Sources 03 - Add Nested Transaction Support to Index Plugins
## Background & Context

We are implementing [Checkpoint-Based Recovery](https://github.com/drasi-project/design-documents/tree/main/drasi-lib/Source-Checkpoints) for at-least-once delivery in Drasi Lib.

A key requirement from [Query Checkpointing (doc 02) — Nested Transactions](https://github.com/drasi-project/design-documents/blob/main/drasi-lib/Source-Checkpoints/02-query-checkpointing.md#nested-transactions-to-avoid-changes-in-core-engine) is that `drasi-lib` must write a source checkpoint into the **same database transaction** that the core engine uses for index updates. This ensures the checkpoint and index state are always consistent — if one rolls back, both roll back.

The core engine's `ContinuousQuery::process_source_change()` internally creates a `SessionGuard` that calls `begin()`/`commit()` on the shared `SessionControl`. To write additional data (the checkpoint) into that same transaction without modifying the core engine, `drasi-lib` needs to wrap core's processing in an **outer** `begin()`/`commit()` pair:

```
Lib:    session_control.begin()        ← opens real DB transaction (depth=1)
Core:     session_control.begin()      ← nested, depth++ only (depth=2)
Core:     ...index writes...
Core:     session_control.commit()     ← nested, depth-- only (depth=1)
Lib:    checkpoint_writer.stage(...)   ← writes into the still-open txn
Lib:    session_control.commit()       ← real commit (depth=0)
```

**Today, both `RocksDbSessionState::begin()` and `GarnetSessionState::begin()` return an error if a transaction is already active.** This issue adds a nesting depth counter so inner calls are no-ops.

### 1. RocksDB: `RocksDbSessionState` (`components/indexes/rocksdb/src/session_state.rs`)

Add an `AtomicU32` depth counter. Modify:

- **`begin()`**: If `active_txn` is `None`, create the real transaction and set depth=1 (current behavior). If `active_txn` is `Some`, increment depth and return `Ok(())` (new: nested no-op).
- **`commit()`**: If depth > 1, decrement depth and return `Ok(())` (nested no-op). If depth == 1, take and commit the real transaction, set depth=0 (current behavior).
- **`rollback()`**: Always take and drop the real transaction regardless of depth (a rollback at any level aborts everything). Reset depth=0.

### 2. Garnet: `GarnetSessionState` (`components/indexes/garnet/src/session_state.rs`)

Same pattern with the `active_buffer`:

- **`begin()`**: If `active_buffer` is `None`, create buffer and set depth=1. If `Some`, increment depth.
- **`commit()`**: If depth > 1, decrement. If depth == 1, drain buffer into pipeline and execute.
- **`rollback()`**: Always drop the buffer. Reset depth=0.

### 3. No changes to `NoOpSessionControl` (`core/src/interface/session_control.rs`)

`NoOpSessionControl` is used for in-memory indexes where checkpointing is disabled. No nesting support is needed — `drasi-lib` will not wrap the processing in an outer transaction for volatile indexes.

### 4. No changes to `SessionGuard` or `SessionControl` trait

The trait interface (`begin`, `commit`, `rollback`) is unchanged. `SessionGuard` works as-is. The nesting is purely an implementation detail inside the persistent backends.

## Testing

For each backend (RocksDB, Garnet):

- **Basic nesting**: outer begin → inner begin → inner commit → writes succeed → outer commit succeeds. Verify all writes are persisted.
- **Nested rollback**: outer begin → inner begin → inner commit → outer rollback. Verify nothing is persisted (real txn was rolled back).
- **Inner rollback aborts all**: outer begin → inner begin → inner rollback. Verify depth resets and the real transaction is aborted.
- **Depth > 2**: begin × 3 → commit × 3. Only the last commit does real work.
- **Existing behavior preserved**: single begin → commit works exactly as before (depth never exceeds 1).

## Implementation Notes

- **No core engine changes.** This is purely index plugin work.
- Use `AtomicU32` with `Ordering::SeqCst` for the depth counter (simplest, no performance concern since transactions are already heavyweight).

### PR DESCRIPTION
**Pull request by amansinghoriginal**
## Summary

- Add depth-counted nesting to `RocksDbSessionState` and `GarnetSessionState` so inner `begin()`/`commit()` calls are no-ops when an outer transaction is already active
- Rollback at any nesting depth aborts the real transaction and resets depth to 0
- No changes to `drasi-core`, the `SessionControl` trait, or `NoOpSessionControl`
- Add 9 new RocksDB session state tests (previously had zero) and 6 new Garnet nesting tests

Foundational plumbing for [checkpoint-based recovery](https://github.com/drasi-project/design-documents/tree/main/drasi-lib/Source-Checkpoints) — enables drasi-lib to wrap `process_source_change()` in an outer transaction and stage a checkpoint atomically with core's index writes.

Closes #361

## Testing done

- `cargo test -p drasi-index-rocksdb` — 17 unit + 39 integration tests pass
- `cargo test -p drasi-index-garnet` — 40 non-Redis tests pass (Redis tests require Docker)
- `cargo clippy -p drasi-index-rocksdb -p drasi-index-garnet --all-targets` — clean
- `cargo build --workspace` — full workspace builds

### CODE REVIEW

File: components/indexes/garnet/src/session_state.rs [hunk 1]
```diff
@@ -416,68 +416,135 @@ impl WriteBuffer {
...
+    fn deref(&self) -> &Self::Target {
+        &self.buffer
+    }
+}
+
+impl std::ops::DerefMut for SessionInner {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.buffer
+    }
+}
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/indexes/garnet/src/session_state.rs]: Good catch. Removed Deref/DerefMut and replaced with narrow inherent methods.

File: components/indexes/garnet/src/session_state.rs [hunk 2]
```diff
@@ -516,11 +596,24 @@ impl GarnetSessionState {
...
         // Use get_mut() since &mut self guarantees exclusive access.
-        if let Ok(inner) = self.active_buffer.get_mut() {
-            if inner.is_some() {
+        if let Ok(inner) = self.inner.get_mut() {
+            debug_assert!(
+                inner.depth == 0,
+                "GarnetSessionState dropped with depth {} — should be 0",
+                inner.depth
+            );
+
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/indexes/garnet/src/session_state.rs]: Agreed. Removed!

File: components/indexes/rocksdb/src/session_state.rs [hunk 3]
```diff
@@ -18,98 +18,162 @@ use async_trait::async_trait;
...
 
+/// Groups the active transaction and nesting depth under a single mutex.
+///
+/// Invariants:
+/// - `depth == 0` ↔ `txn.is_none()` (no active session)
+/// - `depth > 0`  ↔ `txn.is_some()` (session active)
+pub(crate) struct SessionInner {
+    pub(crate) txn: Option<Transaction<'static, OptimisticTransactionDB>>,
+    depth: u32,
+}
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/indexes/rocksdb/src/session_state.rs:32]: Agreed - Made SessionInner, its fields, and lock() all private.

File: components/indexes/rocksdb/src/session_state.rs [hunk 4]
```diff
@@ -18,98 +18,162 @@ use async_trait::async_trait;
...
+            "depth==0 but txn is Some — invariant violated"
+        );
+        let txn = self.db.transaction();
+        // SAFETY: The Arc<OptimisticTransactionDB> guarantees the DB outlives the
+        // transaction. We clear the txn in Drop before the Arc is released.
+        // See the safety comment on the struct.
+        let txn: Transaction<'static, OptimisticTransactionDB> =
+            unsafe { std::mem::transmute(txn) };
+        guard.txn = Some(txn);
+        guard.depth = 1;
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/indexes/rocksdb/src/session_state.rs:110]: Keeping as-is because ContinuousQuery::change_lock serializes all session operations per query, so there's no contention to minimize.

File: components/indexes/rocksdb/src/session_state.rs [hunk 5]
```diff
@@ -134,19 +198,33 @@ impl Drop for RocksDbSessionState {
...
             Err(poisoned) => poisoned.into_inner(),
         };
-        if inner.is_some() {
+
+        debug_assert!(
+            inner.depth == 0,
+            "RocksDbSessionState dropped with depth {} — should be 0",
+            inner.depth
+        );
+
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/indexes/rocksdb/src/session_state.rs]: Agreed and fixed.

**Review by agentofreality — APPROVED** [review]

**Review by danielgerlag — APPROVED** [review]
